consensus/safety-rules/src/persistent_safety_storage.rs

```diff
@@ -18,6 +18,8 @@ pub struct PersistentSafetyStorage {
     internal_store: Box<dyn Storage>,
 }
 
+// @REVIEW: it seems that updates are made on a per key basis. What happens if the process crashes
+// mid update? Is there a need for atomic updates?
```

Comment:
Not sure atomic save would even be possible since that would need to be a primitive on the storage solutions. We could consider a bulk update (i.e., submitting several http requests together), but that is just a mitigation and not a resolution. The ordering does matter though to ensure safety. For example, if you're going to move into the next epoch, you should first set epoch -> next epoch and then set preferred and last voted to 0. If you crash, you're in a bad state but only until you catch up. It might be worth exploring some means to make this appear atomic, set all fields and then have a single value that points to the current set of values.

At the same time, if it does crash, we can log these operations a priori so that the admin could fix safety rules directly. This might be a good case to leverage a human operator.

Reply:
one easy way to ensure atomic is to store a serialized ConsensusState in storage instead of separate fields.

Reply:
Hahaha, that is true, let's explore this especially for the triplet of integers that actually have this dependency.

Reply:
So I recall, I specifically wrote the structure of the code so that any crash would be recoverable into a safe state (except one case I'll mention at the end): https://github.com/libra/libra/blob/master/consensus/safety-rules/src/safety_rules.rs#L251

Now the one case we do not handle is the scenario in which a generated vote is lost. Specifically, prior to signing a vote, we set our last_voted_round, sign the vote, and then return it. If there is a crash in safety rules after storing the value and before sending, or consensus crashes before sending it out, then we could have a liveness issue if there are only 2f + 1 honest nodes and this is considered one of the honest nodes.